### PR TITLE
Add composition overloads

### LINKED ISSUE #125: Implement composition for 0 arity and void returning functions
Currently, the function composition extension methods cannot be invoked against zero arity functions or void returning functions:

```
var readline = fun(Console.ReadLine);
var readlineDummyArg = fun((Unit _) => Console.ReadLine());

readline.BackCompose(File.ReadAllBytes); // Cannot resolve matching overload
readlineDummyArg.BackCompose(File.ReadAllBytes);
```

A full set of extension methods would look like the following:

```
public static class ___ComposeExt
{
    public static Func<TOut> Compose<TMid, TOut>(this Func<TMid> f, Func<TMid, TOut> g) 
        => () => g(f());
    public static Func<TIn, TOut> Compose<TIn, TMid, TOut>(this Func<TIn, TMid> f, Func<TMid, TOut> g) 
        => arg => g(f(arg));

    public static Action Compose<TMid>(this Func<TMid> f, Action<TMid> g) 
        => () => g(f());
    public static Action<TIn> Compose<TIn, TMid>(this Func<TIn, TMid> f, Action<TMid> g) 
        => arg => g(f(arg));
}
```

Since one of the goals of the library is to eliminate `void`, the latter two can be substituted with:

```
    public static class ___ComposeExt
    {
   ...
        public static Func<Unit> Compose<TMid>(this Func<TMid> f, Action<TMid> g)
            => () => { g(f()); return Unit.Default; };
        public static Func<TIn, Unit> Compose<TIn, TMid>(this Func<TIn, TMid> f, Action<TMid> g)
            => arg => { g(f(arg)); return Unit.Default; };
    }
```

I realize my terminology in the snippets is a bit wrong because what I term `Compose` is actually `BackCompose`, but you get the gist. In total we should end up with 4 overloads for `Compose` and 4 overloads for `BackCompose`, plus possibly additional overloads for the `compose` function introduced by `Prelude`.

### PR DESCRIPTION
**Pull request by masaeedu**
Add overloads for zero-arity and void returning functions.

Fixes #125

### CODE REVIEW

**Comment by IDisposable** [comment]
Would love to see some unit-test coverage for these.

**Comment by louthy** [comment]
@masaeedu Hi Asad, could you please provide unit tests for your new functionality.  Thanks,

**Comment by masaeedu** [comment]
@louthy @IDisposable Yeah, working on it. There wasn't any unit tests for the existing `Compose`/`BackCompose` extension methods nor the `Prelude.compose` static overloads; I'll add coverage for the `Compose` and `BackCompose` extension methods, but you probably want to use codegen for methods with lots of overloads like `Prelude.compose`.

**Comment by masaeedu** [comment]
@louthy Done. The CI output looks kind of weird, it's spewing lots of HTML, but I ran the tests locally and they compile & pass.

**Comment by louthy** [comment]
@masaeedu The CI is broken atm, haven't gotten around to fixing it.  I'll grab locally and test.

File: LanguageExt.Core/Compose.cs [hunk 1]
```diff
@@ -12,17 +12,65 @@ public static class __ComposeExt
...
+        [Pure]
+        public static Func<T2> Compose<T1, T2>(this Func<T1, T2> b, Func<T1> a) =>
+            () => b(a());
+
+        /// <summary>
+        /// Action composition
+        /// </summary>
+        /// <returns>v => b(a(v))</returns>
+        [Pure]
+        public static Func<T1, Unit> Compose<T1, T2>(this Action<T2> b, Func<T1, T2> a)
```
Comment (masaeedu) [inline LanguageExt.Core/Compose.cs]: I'm fine with having an `Action` return value, I was just trying to be in line with the library's goal of eliminating void. Note that we can't have it both ways; two overloads can't differ only in return type. Keeping that in mind, should we go ahead and change:

```
public static Func<T1, Unit> Compose<T1, T2>(this Action<T2> b, Func<T1, T2> a)
```

to:

```
public static Action<T1> Compose<T1, T2>(this Action<T2> b, Func<T1, T2> a)
```

?

Comment (masaeedu) [inline LanguageExt.Core/Compose.cs]: Also, you don't need a separate overload for an argument of type `this Func<T2, Unit>`, that's already covered by the existing single argument function composition overloads.

**Comment by masaeedu** [comment]
@louthy ping

**Comment by louthy** [comment]
@masaeedu 

Looks good, thanks Asad

**Pull request merged**